api/src/rpc.c

```diff
@@ -0,0 +1,367 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "api/inc/rpc.h"
+#include "api/inc/rpc_exports.h"
+#include "api/inc/rpc_gateway.h"
+#include "api/inc/vmpu_exports.h"
+#include "api/inc/pool_queue_exports.h"
+#include "api/inc/error.h"
+#include "api/inc/uvisor_semaphore.h"
+#include <string.h>
+
+extern UvisorBoxIndex * __uvisor_ps;
+
+
+static uvisor_pool_queue_t * outgoing_message_queue(void)
+{
+    return __uvisor_ps->rpc_outgoing_message_queue;
+}
+
+static uvisor_rpc_message_t * outgoing_message_array(void)
+{
+    return (uvisor_rpc_message_t *) outgoing_message_queue()->pool.array;
+}
+
+static uvisor_pool_queue_t * incoming_message_queue(void)
+{
+    return __uvisor_ps->rpc_incoming_message_queue;
+}
+
+static uvisor_rpc_message_t * incoming_message_array(void)
+{
+    return (uvisor_rpc_message_t *) incoming_message_queue()->pool.array;
+}
+
+static uvisor_pool_queue_t * outgoing_result_queue(void)
+{
+    return __uvisor_ps->rpc_outgoing_result_queue;
+}
+
+static uvisor_rpc_result_obj_t * outgoing_result_array(void)
+{
+    return (uvisor_rpc_result_obj_t *) outgoing_result_queue()->pool.array;
+}
+
+static uvisor_pool_t * fn_group_pool(void)
+{
+    return __uvisor_ps->rpc_fn_group_pool;
+}
+
+static uvisor_rpc_fn_group_t * fn_group_array(void)
+{
+    return (uvisor_rpc_fn_group_t *) fn_group_pool()->array;
+}
+
+/* Place a message into the outgoing queue. `timeout_ms` is how long to wait
+ * for a slot in the outgoing queue before giving up. `msg_slot` is set to the
+ * slot of the message that was allocated. Returns non-zero on failure. */
+static int send_outgoing_rpc(uint32_t p0, uint32_t p1, uint32_t p2, uint32_t p3, const TFN_Ptr fn, uint32_t timeout_ms,
+                             uvisor_rpc_result_t * cookie)
+{
+    static uint32_t shared_result_counter = 0;
```

Comment:
This shared counter is not in protected memory. Any box can write to it and influence the counter value that another box receives in the cookie (like writing `0` in a `while(1)` loop).